[Gradient Compression] Add a documentation page for DDP communication hooks

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51715 [Gradient Compression] Add a documentation page for DDP communication hooks**

Add a documentation page for DDP communication hooks.

![image](https://user-images.githubusercontent.com/8884619/106973305-79279000-6707-11eb-99ad-c5693b47fda8.png)

Facebook:
Full HTML file: P172254228

Differential Revision: [D26249330](https://our.internmc.facebook.com/intern/diff/D26249330/)

Commit:
[Gradient Compression] Add a documentation page for DDP communication hooks

Add a documentation page for DDP communication hooks.

Differential Revision: [D26249330](https://our.internmc.facebook.com/intern/diff/D26249330/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add a documentation page for DDP communication hooks"


Add a documentation page for DDP communication hooks.

Differential Revision: [D26249330](https://our.internmc.facebook.com/intern/diff/D26249330/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add a documentation page for DDP communication hooks"


Add a documentation page for DDP communication hooks.

Differential Revision: [D26249330](https://our.internmc.facebook.com/intern/diff/D26249330/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add a documentation page for DDP communication hooks"


Add a documentation page for DDP communication hooks.

Differential Revision: [D26249330](https://our.internmc.facebook.com/intern/diff/D26249330/)

[ghstack-poisoned]

rohan-varma commented on docs/source/ddp_comm_hooks.rst:
```diff
@@ -0,0 +1,64 @@
+DDP Communication Hooks
+=======================
+
+DDP communication hook is a generic interface to control how to communicate
```
Might be good to explicitly note that these hooks are used in conjunction with DistributedDataParallel in the first paragraph and note that it will override the allreduce.

SciPioneer replied:
Updated the first sentence as:

DDP communication hook is a generic interface to control how to communicate gradients across workers and override the vanilla allreduce in `DistributedDataParallel`.

rohan-varma commented on docs/source/ddp_comm_hooks.rst:
```diff
@@ -0,0 +1,64 @@
+strategies for more advanced use cases.
+
+.. warning ::
+    DDP communication hooks can only support single process single device mode
```
As a side note, may be good to double check that we throw the appropriate errors if wrong backend or mode is used.

SciPioneer replied:
1) single-process single-device mode is already checked in `reducer.cpp`.
2) will add another check for the backend in a separate PR.

rohan-varma commented on docs/source/ddp_comm_hooks.rst:
```diff
@@ -0,0 +1,64 @@
+To use a communication hook, the user just needs to let the DDP model register
+the hook before the training loop.
+
+.. automethod:: torch.nn.parallel.DistributedDataParallel.register_comm_hook
```
Can you also add a code example for how a user might define an allreduce hook, and then call it with a state object?

pritamdamania87 replied:
I believe the docstrings for `register_comm_hook` do have a couple of examples.

SciPioneer replied:
Yes, already `allreduce_hook` already has an example in the docstrings.

rohan-varma commented on docs/source/ddp_comm_hooks.rst:
```diff
@@ -0,0 +1,64 @@
+.. currentmodule:: torch.distributed.algorithms.ddp_comm_hooks.powerSGD_hook
+.. autoclass:: PowerSGDState
+
+PowerSGD Hooks
```
Maybe Add a note that there are batched and unbatched versions, as well as the tradeoffs between them?

SciPioneer replied:
Already covered by the docstrings. Copied it here:

This variant is faster than powerSGD_hook() that runs layer-wise gradient compression, but it usually results in a much lower accuracy, unless matrix_approximation_rank in the state is 1. Increasing matrix_approximation_rank may not necessarily increase the accuracy, because batching per-parameter tensors without column/row alignment can destroy low-rank structure. Therefore, the user should always consider powerSGD_hook() first, and only consider this variant when a satisfying accuracy can be achieved when matrix_approximation_rank is 1.

rohan-varma reviewed (commented):
Looks good, but can you link/screenshot a page of how the documentation renders currently? Thanks for adding this!

Commit:
Update on "[Gradient Compression] Add a documentation page for DDP communication hooks"


Add a documentation page for DDP communication hooks.

Differential Revision: [D26249330](https://our.internmc.facebook.com/intern/diff/D26249330/)

[ghstack-poisoned]

pritamdamania87 commented on docs/source/ddp_comm_hooks.rst:
```diff
@@ -0,0 +1,64 @@
+DDP Communication Hooks
+=======================
```
We should add a warning at the top of the page mentioning this feature is experimental. Similar to what we have in https://pytorch.org/docs/master/pipeline.html

pritamdamania87 commented on docs/source/ddp_comm_hooks.rst:
```diff
@@ -0,0 +1,64 @@
+in ``register_comm_hook`` is either a process group or ``None``.
+
+.. automodule:: torch.distributed.algorithms.ddp_comm_hooks.default_hooks
+    :members:
```
Just wanted to double check this doesn't pull in the private methods from `default_hooks` right? Also, we probably should not pull in `allreduce_fut` which is a public method (should it be private)?

SciPioneer replied:
`_ allreduce_fut` is a private method, and only `allreduce_hook` and `fp16_compress_hook` are exposed.

pritamdamania87 commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -162,38 +164,48 @@ def maybe_increase_iter(self, bucket):
     but also help abstract away some complexity of the internal optimization of DDP for future communication hook developers.
 
-    TODO(wayi@): The above procedure does two matmul+allreduce steps per iteration --
+    TODO: The above procedure does two matmul+allreduce steps per iteration --
     one left multiplication and one right multiplication.
     For warm-start, can take one such step at a time, and alternate between them.
```
TODOs should not be in the docs, this should probably be a comment in the code that is not visible to the user in docs.

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -1063,9 +1064,6 @@ def register_comm_hook(self, state: object, hook: callable):
             ``register_comm_hook`` API, but it is subject to some subtle differences compared
             to ``torch.futures.Future``.
 
-        .. warning ::
-            DDP communication hook is experimental and subject to change.
-
```
Why was this warning removed? We apply the "experimental and subject to change" warning to all "beta" features and I believe ddp comm hook is still "beta" right?

SciPioneer replied:
I see. Recovered.

Commit:
Update on "[Gradient Compression] Add a documentation page for DDP communication hooks"


Add a documentation page for DDP communication hooks.

Differential Revision: [D26249330](https://our.internmc.facebook.com/intern/diff/D26249330/)

[ghstack-poisoned]

pritamdamania87 commented on docs/source/ddp_comm_hooks.rst:
```diff
@@ -2,12 +2,15 @@ DDP Communication Hooks
 
 DDP communication hook is a generic interface to control how to communicate
-gradients across workers.
+gradients across workers by overriding the vanilla allreduce in ``DistributedDataParallel``.
```
nit: Might be nice to have a clickable link to DDP here: https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel.

SciPioneer commented:
> Looks good, but can you link/screenshot a page of how the documentation renders currently? Thanks for adding this!

![Uploading image.png…]()

Commit:
Update on "[Gradient Compression] Add a documentation page for DDP communication hooks"


Add a documentation page for DDP communication hooks.

![image](https://user-images.githubusercontent.com/8884619/106973305-79279000-6707-11eb-99ad-c5693b47fda8.png)

Facebook:
Full HTML file: P172254228

Differential Revision: [D26249330](https://our.internmc.facebook.com/intern/diff/D26249330/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add a documentation page for DDP communication hooks"


Add a documentation page for DDP communication hooks.

![image](https://user-images.githubusercontent.com/8884619/106973305-79279000-6707-11eb-99ad-c5693b47fda8.png)

Facebook:
Full HTML file: P172254228

Differential Revision: [D26249330](https://our.internmc.facebook.com/intern/diff/D26249330/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Add a documentation page for DDP communication hooks"


Add a documentation page for DDP communication hooks.

![image](https://user-images.githubusercontent.com/8884619/106973305-79279000-6707-11eb-99ad-c5693b47fda8.png)

Facebook:
Full HTML file: P172254228

Differential Revision: [D26249330](https://our.internmc.facebook.com/intern/diff/D26249330/)

[ghstack-poisoned]

ngimel commented:
Reverting, it broke the doc build.
